contracts/extensions/ReputationBootstrapper.sol

```diff
@@ -0,0 +1,137 @@
+
+  // Events
+
+  event GrantSet(bytes32 hashedPin, uint256 reputationAmount);
+  event GrantClaimed(address recipient, uint256 reputationAmount, uint256 tokenAmount);
+
+  // Data structures
+
+  struct Grant {
+    uint256 amount;
+    uint256 timestamp;
+  }
+
+  // Storage
+
+  address public token;
+
+  uint256 decayNumerator;
+  uint256 decayDenominator;
+
+  mapping (bytes32 => Grant) public grants;
+
+  // Modifiers
+
+  modifier onlyRoot() {
+    require(colony.hasUserRole(msgSender(), 1, ColonyDataTypes.ColonyRole.Root), "reputation-bootsrapper-caller-not-root");
+    _;
+  }
+
+  // Overrides
+
+  /// @notice Returns the identifier of the extension
+  function identifier() public override pure returns (bytes32) {
+    return keccak256("ReputationBootstrapper");
+  }
+
+  /// @notice Returns the version of the extension
+  function version() public override pure returns (uint256) {
+    return 1;
+  }
+
+  /// @notice Configures the extension
+  /// @param _colony The colony in which the extension holds permissions
+  function install(address _colony) public override auth {
+    require(address(colony) == address(0x0), "extension-already-installed");
+
+    colony = IColony(_colony);
+    token = colony.getToken();
+
+    address colonyNetwork = colony.getColonyNetwork();
+    address repCycle = IColonyNetwork(colonyNetwork).getReputationMiningCycle(false);
+    (decayNumerator, decayDenominator) = IReputationMiningCycle(repCycle).getDecayConstant();
+  }
+
+  /// @notice Called when upgrading the extension
+  function finishUpgrade() public override auth {}
+
+  /// @notice Called when deprecating (or undeprecating) the extension
+  function deprecate(bool _deprecated) public override auth {
+    deprecated = _deprecated;
+  }
+
+  /// @notice Called when uninstalling the extension
+  function uninstall() public override auth {
+    selfdestruct(address(uint160(address(colony))));
+  }
+
+  // Public
+
+  function setGrants(bytes32[] memory _hashedSecrets, uint256[] memory _amounts) public onlyRoot {
+    require(_hashedSecrets.length == _amounts.length, "reputation-bootsrapper-invalid-arguments");
+
+    for (uint256 i; i < _hashedSecrets.length; i++) {
+      require(_amounts[i] <= INT256_MAX, "repuatation-bootsrapper-invalid-amount");
+      grants[_hashedSecrets[i]] = Grant(_amounts[i], block.timestamp);
+
+      emit GrantSet(_hashedSecrets[i], _amounts[i]);
+    }
+  }
+
+  function claimGrant(uint256 _secret) public {
+    bytes32 hashedPin = keccak256(abi.encodePacked(_secret));
+    uint256 grantAmount = grants[hashedPin].amount;
+    uint256 grantTimestamp = grants[hashedPin].timestamp;
+
+    require(grantAmount > 0, "reputation-bootstrapper-nothing-to-claim");
+
+    delete grants[hashedPin];
+
+    uint256 tokenAmount = min(ERC20(token).balanceOf(address(this)), grantAmount);
+    require(tokenAmount >= uint256(grantAmount) || tokenAmount <= 0, "reputation-bootstrapper-insufficient-tokens");
```

Comment:
It's because of the "optional" token behavior. Basically, users can either give reputation OR reputation and tokens. I chose to implement it implicitly, e.g. without having to initialise with a flag, by simply checking to see if the contract held tokens. If the contract has tokens, send them. If the contract has only enough tokens to partially fill the grant, throw an error. I suppose this was my attempt to avoid introducing a flag, but that would have allowed for a more clear implementation.

Reply:
Ah, I see... I'm not wild about this approach towards that functionality. In particular, the scenario where I've done one batch of bootstrapping and set up a second, if I add the grants before adding the tokens, enthusiastic users won't get the tokens they're expecting.